Ort v1.15 fails to build for i686-pc-windows-msvc targets

Hi,

The title says it all. Under `x86_64-pc-windows-msvc` targets everything works perfectly, but the build fails under `i686-pc-windows-msvc` targets.

<details>

<summary>These are the errors I get</summary>


```bash

    Checking ort v1.15.0
error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\environment.rs:88:56
    |
88  |         let logging_function: sys::OrtLoggingFunction = Some(custom_logger);
    |                                                         ---- ^^^^^^^^^^^^^ expected "C" fn, found "stdcall" fn
    |                                                         |
    |                                                         arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(_, OrtLoggingLevel, _, _, _, _)`
                  found fn item `extern "stdcall" fn(_, OrtLoggingLevel, _, _, _, _) {custom_logger}`
    = note: when the arguments and return types match, functions can be coerced to function pointers
help: the type constructed contains `extern "stdcall" fn(*mut c_void, OrtLoggingLevel, *const i8, *const i8, *const i8, *const i8) {custom_logger}` due to the type of the argument passed
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\environment.rs:88:51
    |
88  |         let logging_function: sys::OrtLoggingFunction = Some(custom_logger);
    |                                                         ^^^^^-------------^
    |                                                              |
    |                                                              this argument influences the type of `Some`
note: tuple variant defined here
   --> C:\Users\user\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\option.rs:572:5
    |
572 |     Some(#[stable(feature = "rust1", since = "1.0.0")] T),
    |     ^^^^

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\environment.rs:99:56
    |
99  |         let logging_function: sys::OrtLoggingFunction = Some(custom_logger);
    |                                                         ---- ^^^^^^^^^^^^^ expected "C" fn, found "stdcall" fn
    |                                                         |
    |                                                         arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(_, OrtLoggingLevel, _, _, _, _)`
                  found fn item `extern "stdcall" fn(_, OrtLoggingLevel, _, _, _, _) {custom_logger}`
    = note: when the arguments and return types match, functions can be coerced to function pointers
help: the type constructed contains `extern "stdcall" fn(*mut c_void, OrtLoggingLevel, *const i8, *const i8, *const i8, *const i8) {custom_logger}` due to the type of the argument passed
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\environment.rs:99:51
    |
99  |         let logging_function: sys::OrtLoggingFunction = Some(custom_logger);
    |                                                         ^^^^^-------------^
    |                                                              |
    |                                                              this argument influences the type of `Some`
note: tuple variant defined here
   --> C:\Users\user\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\option.rs:572:5
    |
572 |     Some(#[stable(feature = "rust1", since = "1.0.0")] T),
    |     ^^^^

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\environment.rs:203:57
    |
203 |             let logging_function: sys::OrtLoggingFunction = Some(custom_logger);
    |                                                             ---- ^^^^^^^^^^^^^ expected "C" fn, found "stdcall" fn
    |                                                             |
    |                                                             arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `unsafe extern "C" fn(_, OrtLoggingLevel, _, _, _, _)`
                  found fn item `extern "stdcall" fn(_, OrtLoggingLevel, _, _, _, _) {custom_logger}`
    = note: when the arguments and return types match, functions can be coerced to function pointers
help: the type constructed contains `extern "stdcall" fn(*mut c_void, OrtLoggingLevel, *const i8, *const i8, *const i8, *const i8) {custom_logger}` due to the type of the argument passed
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\environment.rs:203:52
    |
203 |             let logging_function: sys::OrtLoggingFunction = Some(custom_logger);
    |                                                             ^^^^^-------------^
    |                                                                  |
    |                                                                  this argument influences the type of `Some`
note: tuple variant defined here
   --> C:\Users\user\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\option.rs:572:5
    |
572 |     Some(#[stable(feature = "rust1", since = "1.0.0")] T),
    |     ^^^^

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:778:20
    |
778 |         extract_io_count(f, session_ptr)
    |         ---------------- ^ expected "stdcall" fn, found "C" fn
    |         |
    |         arguments to this function are incorrect
    |
    = note: expected fn pointer `unsafe extern "stdcall" fn(_, _) -> _`
               found fn pointer `unsafe extern "C" fn(_, _) -> _`
note: function defined here
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:786:5
    |
786 |     fn extract_io_count(
    |        ^^^^^^^^^^^^^^^^
787 |         f: extern_system_fn! { unsafe fn(*const sys::OrtSession, *mut usize) -> *mut sys::OrtStatus },
    |         ---------------------------------------------------------------------------------------------

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:783:20
    |
783 |         extract_io_count(f, session_ptr)
    |         ---------------- ^ expected "stdcall" fn, found "C" fn
    |         |
    |         arguments to this function are incorrect
    |
    = note: expected fn pointer `unsafe extern "stdcall" fn(_, _) -> _`
               found fn pointer `unsafe extern "C" fn(_, _) -> _`
note: function defined here
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:786:5
    |
786 |     fn extract_io_count(
    |        ^^^^^^^^^^^^^^^^
787 |         f: extern_system_fn! { unsafe fn(*const sys::OrtSession, *mut usize) -> *mut sys::OrtStatus },
    |         ---------------------------------------------------------------------------------------------

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:802:19
    |
802 |         extract_io_name(f, session_ptr, allocator_ptr, i)
    |         --------------- ^ expected "stdcall" fn, found "C" fn
    |         |
    |         arguments to this function are incorrect
    |
    = note: expected fn pointer `unsafe extern "stdcall" fn(_, _, _, _) -> _`
               found fn pointer `unsafe extern "C" fn(_, _, _, _) -> _`
note: function defined here
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:816:5
    |
816 |       fn extract_io_name(
    |          ^^^^^^^^^^^^^^^
817 | /         f: extern_system_fn! { unsafe fn(
818 | |             *const sys::OrtSession,
819 | |             size_t,
820 | |             *mut sys::OrtAllocator,
821 | |             *mut *mut c_char,
822 | |         ) -> *mut sys::OrtStatus },
    | |__________________________________-

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:807:19
    |
807 |         extract_io_name(f, session_ptr, allocator_ptr, i)
    |         --------------- ^ expected "stdcall" fn, found "C" fn
    |         |
    |         arguments to this function are incorrect
    |
    = note: expected fn pointer `unsafe extern "stdcall" fn(_, _, _, _) -> _`
               found fn pointer `unsafe extern "C" fn(_, _, _, _) -> _`
note: function defined here
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:816:5
    |
816 |       fn extract_io_name(
    |          ^^^^^^^^^^^^^^^
817 | /         f: extern_system_fn! { unsafe fn(
818 | |             *const sys::OrtSession,
819 | |             size_t,
820 | |             *mut sys::OrtAllocator,
821 | |             *mut *mut c_char,
822 | |         ) -> *mut sys::OrtStatus },
    | |__________________________________-

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:839:45
    |
839 |         let (input_type, dimensions) = extract_io(f, session_ptr, i as _)?;
    |                                        ---------- ^ expected "stdcall" fn, found "C" fn
    |                                        |
    |                                        arguments to this function are incorrect
    |
    = note: expected fn pointer `unsafe extern "stdcall" fn(_, _, _) -> _`
               found fn pointer `unsafe extern "C" fn(_, _, _) -> _`
note: function defined here
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:858:5
    |
858 |       fn extract_io(
    |          ^^^^^^^^^^
859 | /         f: extern_system_fn! { unsafe fn(
860 | |             *const sys::OrtSession,
861 | |             size_t,
862 | |             *mut *mut sys::OrtTypeInfo,
863 | |         ) -> *mut sys::OrtStatus },
    | |__________________________________-

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:850:46
    |
850 |         let (output_type, dimensions) = extract_io(f, session_ptr, i as _)?;
    |                                         ---------- ^ expected "stdcall" fn, found "C" fn
    |                                         |
    |                                         arguments to this function are incorrect
    |
    = note: expected fn pointer `unsafe extern "stdcall" fn(_, _, _) -> _`
               found fn pointer `unsafe extern "C" fn(_, _, _) -> _`
note: function defined here
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\session.rs:858:5
    |
858 |       fn extract_io(
    |          ^^^^^^^^^^
859 | /         f: extern_system_fn! { unsafe fn(
860 | |             *const sys::OrtSession,
861 | |             size_t,
862 | |             *mut *mut sys::OrtTypeInfo,
863 | |         ) -> *mut sys::OrtStatus },
    | |__________________________________-

error[E0308]: mismatched types
  --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\lib.rs:97:87
   |
45 | ... ($(#[$meta])* unsafe extern "stdcall" fn $($tt)*);
   |                   ------ expected due to this
...
97 | ...em_fn! { unsafe fn () -> *const ffi::c_char } = (*base).GetVersionString.unwrap();
   |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected "stdcall" fn, found "C" fn
   |
   = note: expected fn pointer `unsafe extern "stdcall" fn() -> _`
              found fn pointer `unsafe extern "C" fn() -> _`

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\registry\src\index.crates.io-6f17d22bba15001f\ort-1.15.0\src\lib.rs:112:78
    |
45  |     ($(#[$meta:meta])* unsafe fn $($tt:tt)*) => ($(#[$meta])* unsafe extern "stdcall" fn $($tt)*);
    |                                                               ------ expected due to this
...
112 |             let get_api: extern_system_fn! { unsafe fn(u32) -> *const sys::OrtApi } = (*base).GetApi.unwrap();
    |                                                                                       ^^^^^^^^^^^^^^^^^^^^^^^ expected "stdcall" fn, found "C" fn
    |
    = note: expected fn pointer `unsafe extern "stdcall" fn(_) -> _`
               found fn pointer `unsafe extern "C" fn(_) -> _`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `ort` (lib) due to 11 previous errors
warning: build failed, waiting for other jobs to finish...

```

</details>

Best
Musharraf

--- a/src/operator/bound.rs
+++ b/src/operator/bound.rs
@@ -9,7 +9,7 @@ use super::{
 	kernel::{Kernel, KernelAttributes, KernelContext},
 	DummyOperator, Operator
 };
-use crate::error::IntoStatus;
+use crate::{error::IntoStatus, extern_system_fn};
 
 #[repr(C)] // <- important! a defined layout allows us to store extra data after the `OrtCustomOp` that we can retrieve later
 pub(crate) struct BoundOperator<O: Operator> {
@@ -65,115 +65,147 @@ impl<O: Operator> BoundOperator<O> {
 		&*op.cast()
 	}
 
-	pub(crate) unsafe extern "C" fn CreateKernelV2(
-		_: *const ort_sys::OrtCustomOp,
-		_: *const ort_sys::OrtApi,
-		info: *const ort_sys::OrtKernelInfo,
-		kernel_ptr: *mut *mut ort_sys::c_void
-	) -> *mut ort_sys::OrtStatus {
-		let kernel = match O::create_kernel(&KernelAttributes::new(info)) {
-			Ok(kernel) => kernel,
-			e => return e.into_status()
-		};
-		*kernel_ptr = (Box::leak(Box::new(kernel)) as *mut O::Kernel).cast();
-		Ok(()).into_status()
+	extern_system_fn! {
+		pub(crate) unsafe fn CreateKernelV2(
+			_: *const ort_sys::OrtCustomOp,
+			_: *const ort_sys::OrtApi,
+			info: *const ort_sys::OrtKernelInfo,
+			kernel_ptr: *mut *mut ort_sys::c_void
+		) -> *mut ort_sys::OrtStatus {
+			let kernel = match O::create_kernel(&KernelAttributes::new(info)) {
+				Ok(kernel) => kernel,
+				e => return e.into_status()
+			};
+			*kernel_ptr = (Box::leak(Box::new(kernel)) as *mut O::Kernel).cast();
+			Ok(()).into_status()
+		}
 	}
 
-	pub(crate) unsafe extern "C" fn ComputeKernelV2(kernel_ptr: *mut ort_sys::c_void, context: *mut ort_sys::OrtKernelContext) -> *mut ort_sys::OrtStatus {
-		let context = KernelContext::new(context);
-		O::Kernel::compute(unsafe { &mut *kernel_ptr.cast::<O::Kernel>() }, &context).into_status()
+	extern_system_fn! {
+		pub(crate) unsafe fn ComputeKernelV2(kernel_ptr: *mut ort_sys::c_void, context: *mut ort_sys::OrtKernelContext) -> *mut ort_sys::OrtStatus {
+			let context = KernelContext::new(context);
+			O::Kernel::compute(unsafe { &mut *kernel_ptr.cast::<O::Kernel>() }, &context).into_status()
+		}
 	}
 
-	pub(crate) unsafe extern "C" fn KernelDestroy(op_kernel: *mut ort_sys::c_void) {
-		drop(Box::from_raw(op_kernel.cast::<O::Kernel>()));
+	extern_system_fn! {
+		pub(crate) unsafe fn KernelDestroy(op_kernel: *mut ort_sys::c_void) {
+			drop(Box::from_raw(op_kernel.cast::<O::Kernel>()));
+		}
 	}
 
-	pub(crate) unsafe extern "C" fn GetName(op: *const ort_sys::OrtCustomOp) -> *const ort_sys::c_char {
-		let safe = Self::safe(op);
-		safe.name.as_ptr()
+	extern_system_fn! {
+		pub(crate) unsafe fn GetName(op: *const ort_sys::OrtCustomOp) -> *const ort_sys::c_char {
+			let safe = Self::safe(op);
+			safe.name.as_ptr()
+		}
 	}
-	pub(crate) unsafe extern "C" fn GetExecutionProviderType(op: *const ort_sys::OrtCustomOp) -> *const ort_sys::c_char {
-		let safe = Self::safe(op);
-		safe.execution_provider_type.as_ref().map(|c| c.as_ptr()).unwrap_or_else(ptr::null)
-	}
-
-	pub(crate) unsafe extern "C" fn GetStartVersion(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
-		O::min_version()
-	}
-	pub(crate) unsafe extern "C" fn GetEndVersion(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
-		O::max_version()
+	extern_system_fn! {
+		pub(crate) unsafe fn GetExecutionProviderType(op: *const ort_sys::OrtCustomOp) -> *const ort_sys::c_char {
+			let safe = Self::safe(op);
+			safe.execution_provider_type.as_ref().map(|c| c.as_ptr()).unwrap_or_else(ptr::null)
+		}
 	}
 
-	pub(crate) unsafe extern "C" fn GetInputMemoryType(_: *const ort_sys::OrtCustomOp, index: ort_sys::size_t) -> ort_sys::OrtMemType {
-		O::inputs()[index as usize].memory_type.into()
+	extern_system_fn! {
+		pub(crate) unsafe fn GetStartVersion(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
+			O::min_version()
+		}
 	}
-	pub(crate) unsafe extern "C" fn GetInputCharacteristic(
-		_: *const ort_sys::OrtCustomOp,
-		index: ort_sys::size_t
-	) -> ort_sys::OrtCustomOpInputOutputCharacteristic {
-		O::inputs()[index as usize].characteristic.into()
-	}
-	pub(crate) unsafe extern "C" fn GetOutputCharacteristic(
-		_: *const ort_sys::OrtCustomOp,
-		index: ort_sys::size_t
-	) -> ort_sys::OrtCustomOpInputOutputCharacteristic {
-		O::outputs()[index as usize].characteristic.into()
-	}
-	pub(crate) unsafe extern "C" fn GetInputTypeCount(_: *const ort_sys::OrtCustomOp) -> ort_sys::size_t {
-		O::inputs().len() as _
-	}
-	pub(crate) unsafe extern "C" fn GetOutputTypeCount(_: *const ort_sys::OrtCustomOp) -> ort_sys::size_t {
-		O::outputs().len() as _
-	}
-	pub(crate) unsafe extern "C" fn GetInputType(_: *const ort_sys::OrtCustomOp, index: ort_sys::size_t) -> ort_sys::ONNXTensorElementDataType {
-		O::inputs()[index as usize]
-			.r#type
-			.map(|c| c.into())
-			.unwrap_or(ort_sys::ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED)
-	}
-	pub(crate) unsafe extern "C" fn GetOutputType(_: *const ort_sys::OrtCustomOp, index: ort_sys::size_t) -> ort_sys::ONNXTensorElementDataType {
-		O::outputs()[index as usize]
-			.r#type
-			.map(|c| c.into())
-			.unwrap_or(ort_sys::ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED)
-	}
-	pub(crate) unsafe extern "C" fn GetVariadicInputMinArity(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
-		O::inputs()
-			.into_iter()
-			.find(|c| c.characteristic == InputOutputCharacteristic::Variadic)
-			.and_then(|c| c.variadic_min_arity)
-			.unwrap_or(1)
-			.try_into()
-			.expect("input minimum arity overflows i32")
-	}
-	pub(crate) unsafe extern "C" fn GetVariadicInputHomogeneity(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
-		O::inputs()
-			.into_iter()
-			.find(|c| c.characteristic == InputOutputCharacteristic::Variadic)
-			.and_then(|c| c.variadic_homogeneity)
-			.unwrap_or(false)
-			.into()
-	}
-	pub(crate) unsafe extern "C" fn GetVariadicOutputMinArity(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
-		O::outputs()
-			.into_iter()
-			.find(|c| c.characteristic == InputOutputCharacteristic::Variadic)
-			.and_then(|c| c.variadic_min_arity)
-			.unwrap_or(1)
-			.try_into()
-			.expect("output minimum arity overflows i32")
-	}
-	pub(crate) unsafe extern "C" fn GetVariadicOutputHomogeneity(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
-		O::outputs()
-			.into_iter()
-			.find(|c| c.characteristic == InputOutputCharacteristic::Variadic)
-			.and_then(|c| c.variadic_homogeneity)
-			.unwrap_or(false)
-			.into()
+	extern_system_fn! {
+		pub(crate) unsafe fn GetEndVersion(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
+			O::max_version()
+		}
 	}
 
-	pub(crate) unsafe extern "C" fn InferOutputShapeFn(_: *const ort_sys::OrtCustomOp, arg1: *mut ort_sys::OrtShapeInferContext) -> *mut ort_sys::OrtStatus {
-		O::get_infer_shape_function().expect("missing infer shape function")(arg1).into_status()
+	extern_system_fn! {
+		pub(crate) unsafe fn GetInputMemoryType(_: *const ort_sys::OrtCustomOp, index: ort_sys::size_t) -> ort_sys::OrtMemType {
+			O::inputs()[index as usize].memory_type.into()
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetInputCharacteristic(_: *const ort_sys::OrtCustomOp, index: ort_sys::size_t) -> ort_sys::OrtCustomOpInputOutputCharacteristic {
+			O::inputs()[index as usize].characteristic.into()
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetOutputCharacteristic(_: *const ort_sys::OrtCustomOp, index: ort_sys::size_t) -> ort_sys::OrtCustomOpInputOutputCharacteristic {
+			O::outputs()[index as usize].characteristic.into()
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetInputTypeCount(_: *const ort_sys::OrtCustomOp) -> ort_sys::size_t {
+			O::inputs().len() as _
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetOutputTypeCount(_: *const ort_sys::OrtCustomOp) -> ort_sys::size_t {
+			O::outputs().len() as _
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetInputType(_: *const ort_sys::OrtCustomOp, index: ort_sys::size_t) -> ort_sys::ONNXTensorElementDataType {
+			O::inputs()[index as usize]
+				.r#type
+				.map(|c| c.into())
+				.unwrap_or(ort_sys::ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED)
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetOutputType(_: *const ort_sys::OrtCustomOp, index: ort_sys::size_t) -> ort_sys::ONNXTensorElementDataType {
+			O::outputs()[index as usize]
+				.r#type
+				.map(|c| c.into())
+				.unwrap_or(ort_sys::ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED)
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetVariadicInputMinArity(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
+			O::inputs()
+				.into_iter()
+				.find(|c| c.characteristic == InputOutputCharacteristic::Variadic)
+				.and_then(|c| c.variadic_min_arity)
+				.unwrap_or(1)
+				.try_into()
+				.expect("input minimum arity overflows i32")
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetVariadicInputHomogeneity(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
+			O::inputs()
+				.into_iter()
+				.find(|c| c.characteristic == InputOutputCharacteristic::Variadic)
+				.and_then(|c| c.variadic_homogeneity)
+				.unwrap_or(false)
+				.into()
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetVariadicOutputMinArity(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
+			O::outputs()
+				.into_iter()
+				.find(|c| c.characteristic == InputOutputCharacteristic::Variadic)
+				.and_then(|c| c.variadic_min_arity)
+				.unwrap_or(1)
+				.try_into()
+				.expect("output minimum arity overflows i32")
+		}
+	}
+	extern_system_fn! {
+		pub(crate) unsafe fn GetVariadicOutputHomogeneity(_: *const ort_sys::OrtCustomOp) -> ort_sys::c_int {
+			O::outputs()
+				.into_iter()
+				.find(|c| c.characteristic == InputOutputCharacteristic::Variadic)
+				.and_then(|c| c.variadic_homogeneity)
+				.unwrap_or(false)
+				.into()
+		}
+	}
+
+	extern_system_fn! {
+		pub(crate) unsafe fn InferOutputShapeFn(_: *const ort_sys::OrtCustomOp, arg1: *mut ort_sys::OrtShapeInferContext) -> *mut ort_sys::OrtStatus {
+			O::get_infer_shape_function().expect("missing infer shape function")(arg1).into_status()
+		}
 	}
 }
 

